starlite/plugins/base.py

```diff
@@ -12,45 +12,258 @@
     Union,
 )
 
-from typing_extensions import Protocol, get_args, runtime_checkable
+from typing_extensions import Protocol, TypeGuard, get_args, runtime_checkable
 
 if TYPE_CHECKING:
     from pydantic import BaseModel
 
+    from starlite.types import (
+        AfterRequestHandler,
+        AfterResponseHandler,
+        BeforeRequestHandler,
+        ControllerRouterHandler,
+        Dependencies,
+        ExceptionHandlersMap,
+        Guard,
+        LifeCycleHandler,
+        Middleware,
+        ParametersMap,
+        ResponseCookies,
+        ResponseHeadersMap,
+        ResponseType,
+    )
+
 ModelT = TypeVar("ModelT")
 
 
 @runtime_checkable
 class PluginProtocol(Protocol[ModelT]):  # pragma: no cover
```

Comment:
presumably, we'd add the additional hooks referenced in #350 here as well?

Reply:
Yup. 

Another alternative is to pass the Starlite instance to the plugin before finishing init. Can make plugins simpler, but  not sure.